Signal/src/ViewControllers/ContactsViewHelper.m

```diff
@@ -277,6 +283,91 @@ - (BOOL)doesContact:(Contact *)contact matchSearchTerms:(NSArray<NSString *> *)s
+                                                            style:UIAlertActionStyleDefault
+                                                          handler:^(UIAlertAction *_Nonnull action) {
+                                                              [[UIApplication sharedApplication] openSystemSettings];
+                                                          }]];
+
+        [fromViewController presentViewController:alertController animated:YES completion:nil];
+        return;
+    }
+
+    CNContactViewController *_Nullable contactViewController;
+    if (signalAccount) {
+        CNContact *_Nullable cnContact = signalAccount.contact.cnContact;
+        if (cnContact) {
+            if (shouldEditImmediately) {
+                // Not acutally a "new" contact, but this brings up the edit form rather than the "Read" form
+                // saving our users a tap in some cases when we already know they want to edit.
+                contactViewController = [CNContactViewController viewControllerForNewContact:cnContact];
+            } else {
+                contactViewController = [CNContactViewController viewControllerForContact:cnContact];
+            }
+        }
+    }
+
+    if (!contactViewController) {
+        CNMutableContact *newContact = [CNMutableContact new];
+        CNPhoneNumber *phoneNumber = [CNPhoneNumber phoneNumberWithStringValue:recipientId];
+        CNLabeledValue<CNPhoneNumber *> *labeledPhoneNumber =
+            [CNLabeledValue labeledValueWithLabel:CNLabelPhoneNumberMain value:phoneNumber];
+        newContact.phoneNumbers = @[ labeledPhoneNumber ];
+
+        contactViewController = [CNContactViewController viewControllerForNewContact:newContact];
+    }
+
+    contactViewController.delegate = fromViewController;
+    contactViewController.allowsActions = NO;
+    contactViewController.allowsEditing = YES;
+    contactViewController.navigationItem.leftBarButtonItem =
+        [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"TXT_CANCEL_TITLE", nil)
+                                         style:UIBarButtonItemStylePlain
+                                        target:fromViewController
+                                        action:@selector(didFinishEditingContact)];
+
+    UINavigationController *navigationController =
+        [[UINavigationController alloc] initWithRootViewController:contactViewController];
+
+    // We want the presentation to imply a "replacement" in this case.
+    if (shouldEditImmediately) {
+        navigationController.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
+    }
+    [fromViewController presentViewController:navigationController animated:YES completion:nil];
+
+    // HACK otherwise CNContactViewController Navbar is shown as black.
+    // RADAR rdar://28433898 http://www.openradar.me/28433898
+    // CNContactViewController incompatible with opaque navigation bar
```

Comment:
Maybe add a note in the header for this method that the caller should call the "opposite" `UIUtil` method in view will appear?